### PR TITLE
Fix CI

### LINKED ISSUE #114: Broken tests on master
There are some broken tests, see [test logs](https://github.com/JuliaGaussianProcesses/TemporalGPs.jl/actions/runs/6236954487/job/16929434915).

### PR DESCRIPTION
**Pull request by simsurace**
This PR is to try to fix #114. I will start by re-testing on Julia 1.8, because the last green CI run was before 1.9 was released.

### CODE REVIEW

**Comment by simsurace** [comment]
1.6 is expected to fail. But interesting that the old manifest + Julia 1.8 are needed to pass all tests.

**Comment by simsurace** [comment]
Ok, adding some `rrule`s was sufficient. They are probably not as fast as they could be and should be upstreamed to `ChainRules.jl`, but it should do for now.
Let's revert the debugging changes now.

**Comment by simsurace** [comment]
Managed to fix it. There was an API problem with `dtc` from AbstractGPs 0.5.16->0.5.17 . I'm now using the new API, I think TemporalGPs should probably not be extending the old one.
Will do some cleanup work later to bump compat and get rid of the deprecation warnings before merging, but I ask for a review to see if the changes are ok.

**Comment by simsurace** [comment]
Unbelieveable, how did it break again?

**Comment by simsurace** [comment]
Seems like I repeated the same mistake as Theo in [this commit](https://github.com/JuliaGaussianProcesses/TemporalGPs.jl/commit/fbd2c01d15b5f5bb864fa3d60e336c61fc35f1cf). Seems like fixing those deprecation warnings is not so easy. I clearly don't know how to fix all this AD stuff :)

**Comment by simsurace** [comment]
The examples seem to take prohibitively long and hitting the timeout.

**Comment by willtebbutt** [comment]
@simsurace modulo the tests timing out, this looks great.

Before I review properly, I'd like to know whether you would you be willing to upstream the `kron` rules to ChainRules.jl? This would avoid type-piracy here, and would ensure that everyone gets the benefit of them.

**Comment by simsurace** [comment]
I meant to do that, but I haven't yet managed to write down rules that pass all tests.

**Comment by willtebbutt** [comment]
Would it be helpful if I took a look at testing the code currently on this branch?

**Comment by simsurace** [comment]
Thanks, I'll open a PR draft and link it here.

**Comment by willtebbutt** [comment]
Okay. I'll take a look at what's going on without the rules in the next hour or two.

**Comment by willtebbutt** [comment]
Sorry, it's going to be tomorrow now. The tests are taking forever (which would explain the time-out). I'm not sure what's happened in 1.9, but the code is _definitely_ slower.

**Comment by willtebbutt** [comment]
Aha! I've figured out why we're not hitting Zygote's `kron` rule. For some reason we're hitting `kron(::AbstractMatrix, ::AbstractVector)`, for which Zygote doesn't have a rule (it only has rules for `kron(::AbstractMatrix, ::AbstractMatrix)` and `kron(::AbstractVector, ::AbstractVector)`.)

Could you please replace the rules with
```julia
Zygote._kron(x::AbstractMatrix, y::AbstractVector) = Zygote._kron(x, reshape(y, :, 1))

function Zygote._pullback(
    cx::Zygote.AContext, ::typeof(kron), a::AbstractMatrix, b::AbstractVector
)
    res, back = Zygote._pullback(cx, Zygote._kron, a, b)
    return res, back ∘ Zygote.unthunk_tangent
end
```
and see if it makes CI pass? It's type-piracy, but I feel like it's a better fix than writing the rule out with for-loops and such.

**Comment by simsurace** [comment]
@willtebbutt yes, tests pass with your Zygote rules.

**Comment by willtebbutt** [comment]
Fantastic. Are you happy to do the following:
1. wait for https://github.com/FluxML/Zygote.jl/pull/1458 to be merged
2. modify this PR to up the lower bound on AbstractGPs and Zygote so that these changes are definitely used by TemporalGPs?

This definitely seems to me like the cleanest solution, but we might have to wait a day or two.

**Comment by simsurace** [comment]
I think the lower bound on AbstractGPs is fine because in this PR I changed the tests to use the new API, so I'm not relying on the deprecated behavior. #117 was just to verify that the new AbstractGPs version would have passed the tests anyway.
So we have two options
1. Keep the PR as is regarding AbstractGPs
2. Bump this PR's compat for AbstractGPs to 0.5.19 and revert the changes in the tests
I'm fine with either one.

**Comment by simsurace** [comment]
On the other hand, the very slow performance still seems to be an issue.
I think the examples CI steps should definitely be able to finish in a reasonable time and be green for this to be merged.

**Comment by simsurace** [comment]
I haven't been able to reproduce the fast behavior. Do you have a manifest from earlier on?

The simple 1d learning takes 40s per iteration on my machine.

**Comment by willtebbutt** [comment]
If you bump the Zygote bound to 0.6.65, you should be able to remove the `kron` rule.

**Comment by simsurace** [comment]
Ok this works! I opened an issue to track the Examples slowness https://github.com/JuliaGaussianProcesses/TemporalGPs.jl/issues/118. I think we can merge this but hold off on the release of 0.6.5.

**Comment by willtebbutt** [comment]
> Ok this works! I opened an issue to track the Examples slowness https://github.com/JuliaGaussianProcesses/TemporalGPs.jl/issues/118. I think we can merge this but hold off on the release of 0.6.5. @willtebbutt are you happy to merge?

Give me a couple of hours -- I think I might have a fix for the performance problems.

**Comment by simsurace** [comment]
I'll be interested in how you debugged this. I profiled the gradient computation and can see a ton of runtime dispatch, but I can't get much beyond that.

**Comment by willtebbutt** [comment]
Okay, I think I know how to fix the remaining performance problem.

It turns out that reverse-mode + `SMatrix` + `Symmetric` interact poorly. To resolve the performance problem, simply replace https://github.com/JuliaGaussianProcesses/TemporalGPs.jl/blob/d32aaa897dea9be30423ce8e0121ed0dd1bce49c/src/models/linear_gaussian_conditionals.jl#L50
with 
```julia
    return Gaussian(A * m + a, (A * P) * A' + Q)
```

I believe that sorting this out also makes this implementation of `ProjectTo`: https://github.com/JuliaGaussianProcesses/TemporalGPs.jl/blob/d32aaa897dea9be30423ce8e0121ed0dd1bce49c/src/util/chainrules.jl#L31-L36 redundant (I do vaguely remember Theo introducing this to get things up and running again, but obviously some performance was sacrificed). If you could just remove it, that would be great, because it's technically type piracy.

If you could make the change, and check that the `exact_time_learning.jl` example works well for you, I would really appreciate it.

In terms of how I reached this conclusion -- I used the VSCode Julia plugin's `@profview` to look at a flame chart of `Zygote.gradient(objective, params)` (taken from the example). It suggested that a _lot_ of time was being spent in the code that Theo wrote (that I've asked you to remove). Initially I thought I should just make that fast, but I later realised that a better solution was just to avoid getting `SMatrix` and `Symmetric` to ever interact. Fortunately, they only really interact in `predict`, and there wasn't a great need to have the call to `symmetric` in it.

**Comment by willtebbutt** [comment]
> So what's up with the numerical stability concerns?

Unclear to me at this point. There's a really good chance that I was being overly cautious when I wrote that. In any case, removing it doesn't appear to break any of the tests. If we find out down the line that there's a problem with removing it, we can consider a different re-write (now that we know the problem, it shouldn't be too hard to work around in one of several ways), and add a breaking test so that it's clear why we need to enforce exact symmetry here.

**Comment by simsurace** [comment]
Let's see how it goes. Is it worth opening an issue on ChainRules.jl to add a rule for this symmetric static array business? I suppose that's going to hit in other places as well?

**Comment by willtebbutt** [comment]
> Let's see how it goes. Is it worth opening an issue on ChainRules.jl to add a rule for this symmetric static array business? I suppose that's going to hit in other places as well?

It might be worth looking into this. The issue is just performance, rather than anything else, so it shouldn't be too hard to handle. It's not something I have time to do right now though.

**Comment by simsurace** [comment]
Yeah agreed, I'm not expecting you to, but since you presumably have a good sense as to what the problem is, maybe it is easier for you to formulate the issue. I think 50x slowdown is still significant.

**Comment by simsurace** [comment]
Unfortunately, the exact_space_time_learning example now fails with
```
ERROR: PosDefException: matrix is not positive definite; Cholesky factorization failed.
Stacktrace:
  [1] checkpositivedefinite
    @ ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/LinearAlgebra/src/factorization.jl:18 [inlined]
  [2] cholesky!(A::LinearAlgebra.Symmetric{Float64, Matrix{Float64}}, ::LinearAlgebra.NoPivot; check::Bool)
    @ LinearAlgebra ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/LinearAlgebra/src/cholesky.jl:268
  [3] cholesky! (repeats 2 times)
    @ ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/LinearAlgebra/src/cholesky.jl:266 [inlined]
  [4] cholesky(A::LinearAlgebra.Symmetric{Float64, Matrix{Float64}}, ::LinearAlgebra.NoPivot; check::Bool)
    @ LinearAlgebra ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/LinearAlgebra/src/cholesky.jl:400
  [5] cholesky (repeats 2 times)
    @ ~/.julia/juliaup/julia-1.9.3+0.x64.linux.gnu/share/julia/stdlib/v1.9/LinearAlgebra/src/cholesky.jl:400 [inlined]
  [6] posterior_and_lml(x::TemporalGPs.Gaussian{Vector{Float64}, Matrix{Float64}}, f::TemporalGPs.SmallOutputLGC{LinearAlgebra.Adjoint{Float64, Matrix{Float64}}, FillArrays.Fill{Float64, 1, Tuple{Base.OneTo{Int64}}}, LinearAlgebra.Diagonal{Float64, Vector{Float64}}}, y::Vector{Float64})
    @ TemporalGPs ~/.julia/dev/TemporalGPs/src/models/linear_gaussian_conditionals.jl:143
  [7] step_logpdf
    @ ~/.julia/dev/TemporalGPs/src/models/lgssm.jl:170 [inlined]
  [8] step_logpdf
    @ ~/.julia/dev/TemporalGPs/src/models/lgssm.jl:166 [inlined]
  [9] scan_emit(f::typeof(TemporalGPs.step_logpdf), xs::Base.Iterators.Zip{Tuple{TemporalGPs.LGSSM{TemporalGPs.GaussMarkovModel{TemporalGPs.Forward, FillArrays.Fill{Matrix{Float64}, 1, Tuple{Base.OneTo{Int64}}}, FillArrays.Fill{Vector{Float64}, 1, Tuple{Base.OneTo{Int64}}}, FillArrays.Fill{Matrix{Float64}, 1, Tuple{Base.OneTo{Int64}}}, TemporalGPs.Gaussian{Vector{Float64}, Matrix{Float64}}}, StructArrays.StructVector{TemporalGPs.SmallOutputLGC{LinearAlgebra.Adjoint{Float64, Matrix{Float64}}, FillArrays.Fill{Float64, 1, Tuple{Base.OneTo{Int64}}}, LinearAlgebra.Diagonal{Float64, Vector{Float64}}}, NamedTuple{(:A, :a, :Q), Tuple{FillArrays.Fill{LinearAlgebra.Adjoint{Float64, Matrix{Float64}}, 1, Tuple{Base.OneTo{Int64}}}, FillArrays.Fill{FillArrays.Fill{Float64, 1, Tuple{Base.OneTo{Int64}}}, 1, Tuple{Base.OneTo{Int64}}}, Vector{LinearAlgebra.Diagonal{Float64, Vector{Float64}}}}}, Int64}}, Vector{Vector{Float64}}}}, state::TemporalGPs.Gaussian{Vector{Float64}, Matrix{Float64}}, idx::UnitRange{Int64})
    @ TemporalGPs ~/.julia/dev/TemporalGPs/src/util/scan.jl:23
 [10] logpdf(model::TemporalGPs.LGSSM{TemporalGPs.GaussMarkovModel{TemporalGPs.Forward, FillArrays.Fill{Matrix{Float64}, 1, Tuple{Base.OneTo{Int64}}}, FillArrays.Fill{Vector{Float64}, 1, Tuple{Base.OneTo{Int64}}}, FillArrays.Fill{Matrix{Float64}, 1, Tuple{Base.OneTo{Int64}}}, TemporalGPs.Gaussian{Vector{Float64}, Matrix{Float64}}}, StructArrays.StructVector{TemporalGPs.SmallOutputLGC{LinearAlgebra.Adjoint{Float64, Matrix{Float64}}, FillArrays.Fill{Float64, 1, Tuple{Base.OneTo{Int64}}}, LinearAlgebra.Diagonal{Float64, Vector{Float64}}}, NamedTuple{(:A, :a, :Q), Tuple{FillArrays.Fill{LinearAlgebra.Adjoint{Float64, Matrix{Float64}}, 1, Tuple{Base.OneTo{Int64}}}, FillArrays.Fill{FillArrays.Fill{Float64, 1, Tuple{Base.OneTo{Int64}}}, 1, Tuple{Base.OneTo{Int64}}}, Vector{LinearAlgebra.Diagonal{Float64, Vector{Float64}}}}}, Int64}}, y::Vector{Vector{Float64}})
    @ TemporalGPs ~/.julia/dev/TemporalGPs/src/models/lgssm.jl:163
 [11] _logpdf
    @ ~/.julia/dev/TemporalGPs/src/gp/lti_sde.jl:67 [inlined]
 [12] logpdf(ft::AbstractGPs.FiniteGP{TemporalGPs.LTISDE{GP{ZeroMean{Float64}, ScaledKernel{Separable{TransformedKernel{SqExponentialKernel{Distances.Euclidean}, ScaleTransform{Float64}}, TransformedKernel{Matern52Kernel{Distances.Euclidean}, ScaleTransform{Float64}}}, Float64}}, ArrayStorage{Float64}}, RectilinearGrid{Float64, Float64, Vector{Float64}, RegularSpacing{Float64}}, LinearAlgebra.Diagonal{Float64, FillArrays.Fill{Float64, 1, Tuple{Base.OneTo{Int64}}}}}, y::Vector{Float64})
    @ TemporalGPs ~/.julia/dev/TemporalGPs/src/gp/lti_sde.jl:60
 [13] objective(params::NamedTuple{(:var_kernel, :λ_space, :λ_time, :var_noise), NTuple{4, Float64}})
    @ Main ~/.julia/dev/TemporalGPs/examples/exact_space_time_learning.jl:52
 [14] call_composed
    @ ./operators.jl:1034 [inlined]
 [15] (::ComposedFunction{typeof(objective), ComposedFunction{typeof(ParameterHandling.value), ParameterHandling.var"#unflatten_to_NamedTuple#17"{Float64, NamedTuple{(:var_kernel, :λ_space, :λ_time, :var_noise), NTuple{4, ParameterHandling.Positive{Float64, typeof(exp), Float64}}}, ParameterHandling.var"#unflatten_to_Tuple#15"{Float64, NTuple{4, Int64}, NTuple{4, Int64}, NTuple{4, ParameterHandling.var"#unflatten_Positive#25"{Float64, ParameterHandling.Positive{Float64, typeof(exp), Float64}, ParameterHandling.var"#unflatten_to_Real#2"{Float64, Float64}}}}}}})(x::Vector{Float64}; kw::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Base ./operators.jl:1031
 [16] (::ComposedFunction{typeof(objective), ComposedFunction{typeof(ParameterHandling.value), ParameterHandling.var"#unflatten_to_NamedTuple#17"{Float64, NamedTuple{(:var_kernel, :λ_space, :λ_time, :var_noise), NTuple{4, ParameterHandling.Positive{Float64, typeof(exp), Float64}}}, ParameterHandling.var"#unflatten_to_Tuple#15"{Float64, NTuple{4, Int64}, NTuple{4, Int64}, NTuple{4, ParameterHandling.var"#unflatten_Positive#25"{Float64, ParameterHandling.Positive{Float64, typeof(exp), Float64}, ParameterHandling.var"#unflatten_to_Real#2"{Float64, Float64}}}}}}})(x::Vector{Float64})
    @ Base ./operators.jl:1031
 [17] value!!(obj::OnceDifferentiable{Float64, Vector{Float64}, Vector{Float64}}, x::Vector{Float64})
    @ NLSolversBase ~/.julia/packages/NLSolversBase/kavn7/src/interface.jl:9
 [18] value!(obj::OnceDifferentiable{Float64, Vector{Float64}, Vector{Float64}}, x::Vector{Float64})
    @ NLSolversBase ~/.julia/packages/NLSolversBase/kavn7/src/interface.jl:28
 [19] value!(obj::Optim.ManifoldObjective{OnceDifferentiable{Float64, Vector{Float64}, Vector{Float64}}}, x::Vector{Float64})
    @ Optim ~/.julia/packages/Optim/Adqv3/src/Manifolds.jl:31
 [20] (::LineSearches.var"#ϕ#1"{Optim.ManifoldObjective{OnceDifferentiable{Float64, Vector{Float64}, Vector{Float64}}}, Vector{Float64}, Vector{Float64}, Vector{Float64}})(α::Float64)
    @ LineSearches ~/.julia/packages/LineSearches/G1LRk/src/LineSearches.jl:25
 [21] (::LineSearches.BackTracking{Float64, Int64})(ϕ::LineSearches.var"#ϕ#1"{Optim.ManifoldObjective{OnceDifferentiable{Float64, Vector{Float64}, Vector{Float64}}}, Vector{Float64}, Vector{Float64}, Vector{Float64}}, αinitial::Float64, ϕ_0::Float64, dϕ_0::Float64)
    @ LineSearches ~/.julia/packages/LineSearches/G1LRk/src/backtracking.jl:114
 [22] BackTracking
    @ ~/.julia/packages/LineSearches/G1LRk/src/backtracking.jl:33 [inlined]
 [23] BackTracking
    @ ~/.julia/packages/LineSearches/G1LRk/src/backtracking.jl:23 [inlined]
 [24] perform_linesearch!(state::Optim.BFGSState{Vector{Float64}, Matrix{Float64}, Float64, Vector{Float64}}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.BackTracking{Float64, Int64}, Nothing, Nothing, Flat}, d::Optim.ManifoldObjective{OnceDifferentiable{Float64, Vector{Float64}, Vector{Float64}}})
    @ Optim ~/.julia/packages/Optim/Adqv3/src/utilities/perform_linesearch.jl:58
 [25] update_state!(d::OnceDifferentiable{Float64, Vector{Float64}, Vector{Float64}}, state::Optim.BFGSState{Vector{Float64}, Matrix{Float64}, Float64, Vector{Float64}}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.BackTracking{Float64, Int64}, Nothing, Nothing, Flat})
    @ Optim ~/.julia/packages/Optim/Adqv3/src/multivariate/solvers/first_order/bfgs.jl:139
 [26] optimize(d::OnceDifferentiable{Float64, Vector{Float64}, Vector{Float64}}, initial_x::Vector{Float64}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.BackTracking{Float64, Int64}, Nothing, Nothing, Flat}, options::Optim.Options{Float64, Nothing}, state::Optim.BFGSState{Vector{Float64}, Matrix{Float64}, Float64, Vector{Float64}})
    @ Optim ~/.julia/packages/Optim/Adqv3/src/multivariate/optimize/optimize.jl:54
 [27] optimize
    @ ~/.julia/packages/Optim/Adqv3/src/multivariate/optimize/optimize.jl:36 [inlined]
 [28] optimize(f::Function, g::Function, initial_x::Vector{Float64}, method::BFGS{LineSearches.InitialStatic{Float64}, LineSearches.BackTracking{Float64, Int64}, Nothing, Nothing, Flat}, options::Optim.Options{Float64, Nothing}; inplace::Bool, autodiff::Symbol)
    @ Optim ~/.julia/packages/Optim/Adqv3/src/multivariate/optimize/interface.jl:156
 [29] top-level scope
    @ ~/.julia/dev/TemporalGPs/examples/exact_space_time_learning.jl:56
```
Could that be related to the change?

**Comment by willtebbutt** [comment]
Interestingly, I'm not seeing this when I run the tests. Which julia version are you running, and have you tried upgrading the Manifest for the examples directory?

**Comment by willtebbutt** [comment]
Additionally, it looks like the examples are still taking _ages_. I wonder whether we shouldn't make the space-time examples a little bit smaller?

**Comment by simsurace** [comment]
> Additionally, it looks like the examples are still taking _ages_. I wonder whether we shouldn't make the space-time examples a little bit smaller?

I think it makes sense to make them smaller for the tests (we can choose N conditionally smaller when in the test environment), but I'm still wondering if they should run for so long. I suppose that you did not choose the examples such that they would take _hours_ to run.

**Comment by simsurace** [comment]
> Interestingly, I'm not seeing this when I run the tests. Which julia version are you running, and have you tried upgrading the Manifest for the examples directory?

I tried once more and can't reproduce the error.

**Comment by simsurace** [comment]
On my machine, a single gradient evaluation in the approx_space_time_learning example takes approx. 300 seconds. If I reduce N to 100 it takes 3 seconds. Is that the expected scaling behavior?

**Comment by willtebbutt** [comment]
Hmm yeah -- that does seem suspicious. I'll have a dig today. Thanks for continuing to dig into this -- having this sorted out will be very good.

**Comment by willtebbutt** [comment]
Finally gotten around to looking into this. I was immediately able to replicate what you're seeing. It definitely looks like an AD issue, because the primal evaluation is still fast.

It's probably something silly where we're hitting a generic fallback or we need to implement `ProjectTo` carefully (which is annoying) for some reason. I'm starting to dig now

**Comment by simsurace** [comment]
I tried to glean something from the profile, but in contrast to the other case there wasn't much runtime dispatch. Maybe you can see something there that I missed.

**Comment by willtebbutt** [comment]
It _looks_ like it's spending all of it's time in `_project` / `ProjectTo`. This shouldn't be happening, because there's no projection work to be done really. I probably just need to figure out how to get it to stop doing redundant work and it'll be fine.

**Comment by simsurace** [comment]
Do you still have the code to produce the figure in the README (the one showing the scaling)? I would be interested to run these benchmarks once this is fixed.

**Comment by willtebbutt** [comment]
I _think_ it's just https://github.com/JuliaGaussianProcesses/TemporalGPs.jl/blob/master/bench/single_output_gps.jl ? In any case, it's really computing the logpdf and its gradient as a function of the kernel parameters and observations for the three Matern kernels that we've got implemented, and a variety of different numbers of observations (definitely uses `RegularSpacing` rather than arbitrary spacing of inputs).

**Comment by simsurace** [comment]
Two-argument `map` strikes back. I think they can be replaced by broadcasts or comprehensions...

**Comment by simsurace** [comment]
Taken together, these optimizations amount to a 2-3 orders of magnitude improvement in approx_space_time_learning for T = 1000.

File: src/space_time/regular_in_time.jl [hunk 1]
```diff
@@ -24,7 +24,16 @@ function Base.collect(x::RegularInTime)
     return [(x, t) for (x, t) in zip(space_inputs, time_inputs)]
 end
 
-Base.getindex(x::RegularInTime, n::Int) = collect(x)[n]
+function Base.getindex(x::RegularInTime, n::Int)
+    n < 0 || throw(BoundsError(x, n))
```
Comment (simsurace) [inline src/space_time/regular_in_time.jl:28]:
```suggestion
    n < 0 && throw(BoundsError(x, n))
```

**Comment by simsurace** [comment]
Finally all green :tada:. I will now restore the original problem sizes for the space-time examples.

@willtebbutt, are you happy to merge this now?

File: src/models/linear_gaussian_conditionals.jl [hunk 2]
```diff
@@ -46,7 +46,9 @@ be equivalent to
 function predict(x::Gaussian, f::AbstractLGC)
     A, a, Q = get_fields(f)
     m, P = get_fields(x)
-    # Symmetric wrapper needed for numerical stability. Do not unwrap.
+    # Previously, there was a symmetric wrapper around `P` in the following line.
+    # It was removed to solve a performance problem with `Symmetric` around static
+    # arrays, with unclear implications on numerical stability.
```
Comment (simsurace) [inline src/models/linear_gaussian_conditionals.jl:51]:
```suggestion
```

**Comment by simsurace** [comment]
Fixes #118

**Comment by simsurace** [comment]
I removed the "needs version bump" label because the version that is currently on master hasn't been released yet.

**Comment by simsurace** [comment]
The SIGTERM in the last example is new.. I will investigate

**Comment by simsurace** [comment]
Ok, it seems that the augmented_inference example is a bit flaky in CI. Locally it has not failed me. I don't know why it occasionally gets killed. @willtebbutt can we merge?

**Review by willtebbutt — APPROVED** [review]
Thanks for sorting this. Yup, please feel to merge!

**Pull request merged**